File: Main2_74.java
import java.util.*;

public class Main2_74 {

    static int[] preorder;
    static int[] inorder;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();

        preorder = new int[n];
        inorder = new int[n];

        for (int i = 0; i < n; i++) {
            preorder[i] = sc.nextInt();
        }

        for (int i = 0; i < n; i++) {
            inorder[i] = sc.nextInt();
        }

        build(0, n - 1, 0, n - 1);

        System.out.println(sb.toString().trim());

        sc.close();
    }

    static void build(int preStart, int preEnd, int inStart, int inEnd) {

        if (preStart > preEnd || inStart > inEnd)
            return;

        int root = preorder[preStart];

        int rootIndex = -1;

        for (int i = inStart; i <= inEnd; i++) {
            if (inorder[i] == root) {
                rootIndex = i;
                break;
            }
        }

        int leftSize = rootIndex - inStart;

        build(preStart + 1,
                preStart + leftSize,
                inStart,
                rootIndex - 1);

        build(preStart + leftSize + 1,
                preEnd,
                rootIndex + 1,
                inEnd);

        sb.append(root).append(" ");
    }
}

File: Main2_74.py
import sys
sys.setrecursionlimit(10**7)

def build(preStart, preEnd, inStart, inEnd):
    if preStart > preEnd or inStart > inEnd:
        return

    root = preorder[preStart]

    rootIndex = -1
    for i in range(inStart, inEnd + 1):
        if inorder[i] == root:
            rootIndex = i
            break

    leftSize = rootIndex - inStart

    build(preStart + 1,
          preStart + leftSize,
          inStart,
          rootIndex - 1)

    build(preStart + leftSize + 1,
          preEnd,
          rootIndex + 1,
          inEnd)

    postorder.append(root)


def main():
    input = sys.stdin.readline

    n = int(input())

    global preorder, inorder, postorder
    preorder = list(map(int, input().split()))
    inorder = list(map(int, input().split()))
    postorder = []

    build(0, n - 1, 0, n - 1)

    print(" ".join(map(str, postorder)))


if __name__ == "__main__":
    main()